File: desktop/src/renderer/webflow-connect/webflow-connect.ts
/**
 * Webflow add-target renderer.
 *
 * Mirrors the Mac WebflowConnect.swift form and the wordpress-connect page.
 * Differences from WordPress:
 *  - Auth is a single bearer token (a pasted Site API token, or an OAuth token
 *    pre-filled by the deep-link handler).
 *  - Content kinds are DYNAMIC — the site's CMS collections. They can't be
 *    hard-coded, so after a successful Test Connection we fetch them via
 *    `webflow.kinds(...)` and render a checkbox per collection.
 *
 * Test Connection must succeed before Connect is enabled (and before kinds can
 * load). On Connect, IPC saves the target and restarts the daemon supervisor.
 */

import type { ContentKind } from '../preload-types.js';

const $ = (id: string) => document.getElementById(id)!;

const labelInput = $('wf-label') as HTMLInputElement;
const siteIdInput = $('wf-site-id') as HTMLInputElement;
const apiTokenInput = $('wf-api-token') as HTMLInputElement;
const kindsGroup = $('wf-kinds');
const kindsHelp = document.querySelector('.kinds-help') as HTMLElement | null;
const testBtn = $('btn-test') as HTMLButtonElement;
const connectBtn = $('btn-connect') as HTMLButtonElement;
const cancelBtn = $('btn-cancel') as HTMLButtonElement;
const testSpinner = $('test-spinner');
const testResult = $('test-result');
const saveError = $('save-error');
const helpLink = $('wf-help-link') as HTMLAnchorElement;
const oauthBtn = $('btn-oauth') as HTMLButtonElement;
const oauthBlock = $('wf-oauth-block');

const DEFAULT_OAUTH_BASE_URL = 'https://spectersync.com';

/** Origin of the OAuth broker: configured `oauthBaseUrl`, else hosted default. */
async function oauthStartUrl(): Promise<string> {
  const cfg = await window.api.config.read().catch(() => null);
  const configured = cfg?.oauthBaseUrl?.trim();
  const base = configured
    ? (configured.endsWith('/') ? configured.slice(0, -1) : configured)
    : DEFAULT_OAUTH_BASE_URL;
  return `${base}/api/oauth/webflow/start`;
}

oauthBtn.addEventListener('click', async () => {
  oauthBtn.disabled = true;
  try {
    await window.api.shell.openExternal(await oauthStartUrl());
  } finally {
    oauthBtn.disabled = false;
  }
});

let testPassed = false;
let editingHandle: string | null = null;
// Collections discovered for the tested site. Each is a `webflow:<slug>` kind.
let availableKinds: ContentKind[] = [];
let preselectedKinds: ContentKind[] = [];

function kindLabel(kind: ContentKind): string {
  // 'webflow:blog-posts' → 'blog-posts'
  return String(kind).replace(/^webflow:/, '');
}

function renderKinds(): void {
  kindsGroup.innerHTML = '';
  if (availableKinds.length === 0) return;
  availableKinds.forEach((kind) => {
    const id = `wf-kind-${kindLabel(kind)}`;
    const wrap = document.createElement('label');
    wrap.className = 'kind-option';
    const cb = document.createElement('input');
    cb.type = 'checkbox';
    cb.id = id;
    cb.value = kind;
    cb.checked = preselectedKinds.includes(kind);
    const span = document.createElement('span');
    span.textContent = kindLabel(kind);
    wrap.append(cb, span);
    kindsGroup.appendChild(wrap);
  });
}

function selectedKinds(): ContentKind[] {
  return availableKinds.filter((kind) => {
    const cb = document.getElementById(`wf-kind-${kindLabel(kind)}`) as HTMLInputElement | null;
    return !!cb?.checked;
  });
}

function updateButtons(): void {
  const hasFields = !!siteIdInput.value.trim() && !!apiTokenInput.value.trim();
  testBtn.disabled = !hasFields;
  // In edit mode the connection is already known-good, so a label/kind change
  // can be saved without re-testing.
  connectBtn.disabled = !(testPassed || editingHandle !== null) || !hasFields;
}

function clearTestResult(): void {
  testResult.textContent = '';
  testResult.classList.remove('ok', 'fail');
}

function resetTestPassed(): void {
  testPassed = false;
  clearTestResult();
  updateButtons();
}

[siteIdInput, apiTokenInput].forEach((el) => {
  el.addEventListener('input', resetTestPassed);
});

helpLink.addEventListener('click', (ev) => {
  ev.preventDefault();
  const url = 'https://developers.webflow.com/data/reference/authentication';
  navigator.clipboard?.writeText(url).catch(() => { /* ignore */ });
  helpLink.textContent = 'Link copied to clipboard';
  setTimeout(() => {
    helpLink.textContent = 'How to create a Webflow Site API token';
  }, 2000);
});

async function loadKinds(siteId: string, apiToken: string): Promise<void> {
  try {
    const res = await window.api.webflow.kinds(siteId, apiToken);
    if (res.ok && res.kinds) {
      availableKinds = res.kinds;
      if (kindsHelp) {
        kindsHelp.textContent = availableKinds.length
          ? 'Choose which collections to sync.'
          : 'No CMS collections found on this site.';
      }
    }
  } catch {
    /* leave kinds empty; user can retry by testing again */
  }
  renderKinds();
}

testBtn.addEventListener('click', async () => {
  const siteId = siteIdInput.value.trim();
  const apiToken = apiTokenInput.value.trim();

  testBtn.disabled = true;
  testSpinner.classList.remove('hidden');
  clearTestResult();

  try {
    const result = await window.api.webflow.test(siteId, apiToken);
    testSpinner.classList.add('hidden');
    if (result.ok) {
      testPassed = true;
      testResult.textContent = result.message ?? 'Connected.';
      testResult.classList.add('ok');
      await loadKinds(siteId, apiToken);
    } else {
      testPassed = false;
      testResult.textContent = result.error ?? result.message ?? 'Connection failed.';
      testResult.classList.add('fail');
    }
  } catch (err) {
    testSpinner.classList.add('hidden');
    testPassed = false;
    testResult.textContent = (err as Error).message;
    testResult.classList.add('fail');
  } finally {
    updateButtons();
  }
});

connectBtn.addEventListener('click', async () => {
  saveError.classList.add('hidden');
  saveError.textContent = '';
  const siteId = siteIdInput.value.trim();
  const apiToken = apiTokenInput.value.trim();
  const label = labelInput.value.trim() || undefined;
  connectBtn.disabled = true;

  try {
    const result = await window.api.webflow.connect(
      siteId,
      { apiToken },
      label,
      selectedKinds(),
    );
    if (result.ok) {
      window.close();
    } else {
      saveError.textContent = result.error ?? 'Could not save Webflow target.';
      saveError.classList.remove('hidden');
      connectBtn.disabled = false;
    }
  } catch (err) {
    saveError.textContent = (err as Error).message;
    saveError.classList.remove('hidden');
    connectBtn.disabled = false;
  }
});

cancelBtn.addEventListener('click', () => {
  window.close();
});

// Pre-fill on load: editing an existing site, or completing an OAuth flow
// (the deep-link handler stashes the OAuth token in `apiToken`).
async function init(): Promise<void> {
  try {
    const pending = await window.api.connect.pending();
    if (pending && pending.platform === 'webflow') {
      editingHandle = pending.handle ?? null;
      if (pending.siteId) siteIdInput.value = pending.siteId;
      if (pending.apiToken) apiTokenInput.value = pending.apiToken;
      if (pending.label) labelInput.value = pending.label;
      if (pending.contentKinds) preselectedKinds = pending.contentKinds;
    }
    // The OAuth launcher only makes sense for a fresh add. When editing an
    // existing site (or finishing an OAuth flow that already filled the token),
    // hide it so it doesn't restart authorization.
    if (editingHandle !== null || apiTokenInput.value.trim()) {
      oauthBlock.classList.add('hidden');
    }
  } catch {
    /* fresh add flow */
  }
  updateButtons();
}

void init();

export {};

File: mac/Sources/Specter/WebflowConnect.swift
import AppKit
import Foundation
import SwiftUI

/// Standalone window for adding a Webflow site to `targets[]`.
///
/// Differs from the WordPress form in two ways:
///  - Auth is a single bearer token (a pasted Site API token, or an OAuth token
///    pre-filled by the deep-link handler).
///  - Content kinds are DYNAMIC — the site's CMS collections. They can't be
///    hard-coded, so after a successful Test Connection the form fetches them via
///    the daemon `kinds` command and renders a checkbox per collection.
@MainActor
final class WebflowConnectController: ObservableObject {
    @Published var label: String = ""
    @Published var siteId: String = ""
    @Published var apiToken: String = ""
    @Published var contentKinds: [String] = []
    /// Live collections discovered for the tested site (`webflow:<slug>`).
    @Published var availableKinds: [String] = []
    @Published var testResult: OnboardingController.TestResult = .untested
    @Published var isTesting = false
    @Published var saveError: String?

    @Published var editingHandle: String?

    /// Credentials as loaded for an edit. Used to tell a label/content-kind-only
    /// edit (creds untouched → already validated, save without re-test) from a
    /// credential change (must re-test).
    private var loadedSiteId = ""
    private var loadedApiToken = ""

    var isEditing: Bool { editingHandle != nil }

    var canTest: Bool {
        !siteId.trimmingCharacters(in: .whitespaces).isEmpty &&
        !apiToken.trimmingCharacters(in: .whitespaces).isEmpty &&
        !isTesting
    }

    /// True when editing an existing target whose credentials are unchanged from
    /// what was loaded — already validated when added, so a label or content-kind
    /// tweak shouldn't demand a fresh Test (which also re-fetches collections).
    private var isUnchangedCredentialEdit: Bool {
        isEditing &&
        siteId.trimmingCharacters(in: .whitespacesAndNewlines) == loadedSiteId &&
        apiToken.trimmingCharacters(in: .whitespacesAndNewlines) == loadedApiToken
    }

    var canSave: Bool {
        if case .ok = testResult { return true }
        // Editing an already-connected site: allow saving label / content-kind
        // changes without forcing a re-test. Changing the site id or token flips
        // this false (and onChange resets testResult), so credential edits still
        // require a successful Test before Save re-enables.
        return isUnchangedCredentialEdit
    }

    func reset() {
        label = ""
        siteId = ""
        apiToken = ""
        contentKinds = []
        availableKinds = []
        testResult = .untested
        saveError = nil
        isTesting = false
        editingHandle = nil
        loadedSiteId = ""
        loadedApiToken = ""
    }

    /// Pre-fill from an existing Webflow target for the Edit flow.
    func loadForEditing(_ target: TargetConfig) {
        guard case .webflow(let wf) = target.adapter else { return }
        editingHandle = target.handle
        label = target.label
        siteId = wf.siteId
        apiToken = wf.apiToken ?? wf.accessToken ?? ""
        loadedSiteId = wf.siteId.trimmingCharacters(in: .whitespacesAndNewlines)
        loadedApiToken = (wf.apiToken ?? wf.accessToken ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        contentKinds = target.contentKinds
        availableKinds = target.contentKinds   // seed; refreshed on Test
        testResult = .untested
        saveError = nil
        isTesting = false
    }

    /// Pre-fill from a completed OAuth flow: the token is known, the site is not.
    func loadFromOAuth(token: String) {
        reset()
        apiToken = token
    }

    func runTest() {
        let site = siteId.trimmingCharacters(in: .whitespacesAndNewlines)
        let token = apiToken.trimmingCharacters(in: .whitespacesAndNewlines)
        isTesting = true
        testResult = .untested
        DispatchQueue.global().async {
            let result = ConnectionTester.runWebflow(siteId: site, apiToken: token)
            // On success, also pull the live collection list for the kind picker.
            var kinds: [String]? = nil
            if case .ok = result {
                kinds = ConnectionTester.runWebflowKinds(siteId: site, apiToken: token)
            }
            DispatchQueue.main.async {
                self.isTesting = false
                self.testResult = result
                if let kinds { self.availableKinds = kinds }
            }
        }
    }

    func save(completion: @escaping (Bool) -> Void) {
        let site = siteId.trimmingCharacters(in: .whitespacesAndNewlines)
        let token = apiToken.trimmingCharacters(in: .whitespacesAndNewlines)
        // Only persist kinds the tested site actually exposes.
        let kinds = contentKinds.filter { availableKinds.contains($0) }
        do {
            try ConfigStore.upsertWebflowTarget(
                siteId: site,
                apiToken: token,
                accessToken: nil,
                contentKinds: kinds,
                label: label,
                editingHandle: editingHandle
            )
            saveError = nil
            completion(true)
        } catch {
            saveError = error.localizedDescription
            completion(false)
        }
    }
}

struct WebflowConnectView: View {
    @ObservedObject var controller: WebflowConnectController
    var onSave: () -> Void
    var onCancel: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            Text(controller.isEditing ? "Edit Webflow site" : "Add a Webflow site")
                .font(.largeTitle).bold()
                .padding(.horizontal, 24)
                .padding(.top, 24)
            Text("Paste a Site API token, or finish an OAuth connection. Specter syncs your CMS collections to markdown.")
                .foregroundStyle(.secondary)
                .padding(.horizontal, 24)
                .padding(.top, 4)
                .padding(.bottom, 18)

            Form {
                if !controller.isEditing {
                    Section {
                        Button {
                            if let url = OAuthController.endpointURL("/api/oauth/webflow/start") {
                                OAuthController.shared.startInApp(url)
                            }
                        } label: {
                            Label("Connect with Webflow", systemImage: "link")
                        }
                        Text("Authorize Webflow in a secure window — Specter fills in the token and site for you. Or paste a Site API token below instead.")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    } header: {
                        Text("Recommended")
                    }
                }

                Section("Site") {
                    TextField("Label (optional)", text: $controller.label,
                              prompt: Text("e.g. My Webflow site"))
                    TextField("Site ID", text: $controller.siteId,
                              prompt: Text("Webflow site ID"))
                        .onChange(of: controller.siteId) { _ in
                            controller.testResult = .untested
                        }
                    SecureField("API token", text: $controller.apiToken)
                        .onChange(of: controller.apiToken) { _ in
                            controller.testResult = .untested
                        }
                    Button {
                        if let url = URL(string: "https://developers.webflow.com/data/reference/authentication") {
                            NSWorkspace.shared.open(url)
                        }
                    } label: {
                        Label("How to create a Webflow Site API token",
                              systemImage: "questionmark.circle")
                            .font(.caption)
                    }
                    .buttonStyle(.link)

                    HStack(spacing: 10) {
                        Button {
                            controller.runTest()
                        } label: {
                            if controller.isTesting {
                                ProgressView().controlSize(.small)
                                Text("Testing…")
                            } else {
                                Label("Test Connection", systemImage: "checkmark.shield")
                            }
                        }
                        .disabled(!controller.canTest)

                        testResultView
                    }
                }

                Section {
                    if controller.availableKinds.isEmpty {
                        Text("Run Test Connection to load this site's CMS collections.")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    } else {
                        ContentKindSelector(platform: .webflow,
                                            selected: $controller.contentKinds,
                                            availableKinds: controller.availableKinds)
                    }
                } header: {
                    Text("Collections to sync")
                } footer: {
                    Text("Choose which collections to sync. Both directions — only ticked collections pull from and push to Webflow.")
                        .font(.caption)
                        .foregroundStyle(.secondary)
                }
            }
            .formStyle(.grouped)
            .padding(.horizontal, 14)

            if let saveError = controller.saveError {
                Label(saveError, systemImage: "xmark.octagon.fill")
                    .font(.caption)
                    .foregroundStyle(.red)
                    .padding(.horizontal, 24)
                    .padding(.bottom, 8)
            }

            Divider()
            HStack {
                Spacer()
                Button("Cancel", action: onCancel)
                Button(controller.isEditing ? "Save" : "Connect") {
                    controller.save { success in
                        if success { onSave() }
                    }
                }
                .disabled(!controller.canSave)
                .keyboardShortcut(.defaultAction)
            }
            .padding(16)
        }
        .frame(width: 560, height: 480)
    }

    @ViewBuilder
    private var testResultView: some View {
        switch controller.testResult {
        case .untested:
            EmptyView()
        case .ok(let msg):
            Label(msg, systemImage: "checkmark.circle.fill")
                .foregroundStyle(.green)
                .lineLimit(2)
        case .failed(let msg):
            Label(msg, systemImage: "xmark.octagon.fill")
                .foregroundStyle(.red)
                .lineLimit(2)
        }
    }
}
